## Importing the libraries

In [1]:
import pandas as pd

In [2]:
data=pd.read_csv(r"C:\Users\Nikhitha Siva\Desktop\Connect_Mobile__Attrition_Data_file.csv") #loading the data

In [3]:
data.columns

Index(['active_cust', 'income', 'expenditure', 'months_on_network',
       'Num_complaints', 'number_plan_changes', 'relocated', 'monthly_bill',
       'technical_issues_per_month', 'Speed_test_result'],
      dtype='object')

In [4]:
data.isnull().any() #checking for any missing values

active_cust                   False
income                        False
expenditure                   False
months_on_network             False
Num_complaints                False
number_plan_changes           False
relocated                     False
monthly_bill                  False
technical_issues_per_month    False
Speed_test_result             False
dtype: bool

In [5]:
data.isnull().sum()

active_cust                   0
income                        0
expenditure                   0
months_on_network             0
Num_complaints                0
number_plan_changes           0
relocated                     0
monthly_bill                  0
technical_issues_per_month    0
Speed_test_result             0
dtype: int64

In [6]:
data.info #information about the data

<bound method DataFrame.info of        active_cust  income  expenditure  months_on_network  Num_complaints  \
0                0    1586       1268.8                 85               4   
1                0    1581       1264.8                 85               4   
2                0    1594       1275.2                 82               4   
3                0    1594       1275.2                 82               4   
4                1    1609       1287.2                 80               4   
...            ...     ...          ...                ...             ...   
99995            1    1673       1338.4                 71               4   
99996            1    1667       1333.6                 71               4   
99997            1    1672       1337.6                 71               4   
99998            1    1686       1348.8                 69               4   
99999            1    1687       1349.6                 69               4   

       number_plan_changes  rel

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 10 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   active_cust                 100000 non-null  int64  
 1   income                      100000 non-null  int64  
 2   expenditure                 100000 non-null  float64
 3   months_on_network           100000 non-null  int64  
 4   Num_complaints              100000 non-null  int64  
 5   number_plan_changes         100000 non-null  int64  
 6   relocated                   100000 non-null  int64  
 7   monthly_bill                100000 non-null  int64  
 8   technical_issues_per_month  100000 non-null  int64  
 9   Speed_test_result           100000 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 7.6 MB


In [8]:
data.shape #checking the number of rows and columns in the data

(100000, 10)

In [9]:
data.describe()

active_cust         income    expenditure  months_on_network  \
count  100000.000000  100000.000000  100000.000000      100000.000000   
mean        0.578590    1855.279910    1409.669662          58.402450   
std         0.493787    2712.385257    2174.330870          18.643736   
min         0.000000    1130.000000     904.000000         -20.000000   
25%         0.000000    1674.000000    1248.650000          47.000000   
50%         1.000000    1805.000000    1358.400000          56.000000   
75%         1.000000    1907.000000    1494.400000          70.000000   
max         1.000000  229821.000000  183856.800000         449.000000   

       Num_complaints  number_plan_changes      relocated   monthly_bill  \
count   100000.000000        100000.000000  100000.000000  100000.000000   
mean         4.012840             1.385410       0.123480      95.816450   
std          0.285265             1.617101       0.328989      50.161327   
min          3.000000             0.000000       0.000000      30.000000   
25%          4.000000             0.000000       0.000000      45.000000   
50%          4.000000             1.000000       0.000000      87.000000   
75%          4.000000             2.000000       0.000000     133.000000   
max          6.000000             9.000000       1.000000     181.000000   

       technical_issues_per_month  Speed_test_result  
count                100000.00000      100000.000000  
mean                      3.31673          81.668540  
std                       1.24147           9.952554  
min                       1.00000          72.000000  
25%                       3.00000          78.000000  
50%                       4.00000          81.000000  
75%                       4.00000          85.000000  
max                       7.00000         332.000000

## Splitting the Dependent and Independent Variables

In [10]:
X=data.iloc[:,1:9]
X

income  expenditure  months_on_network  Num_complaints  \
0        1586       1268.8                 85               4   
1        1581       1264.8                 85               4   
2        1594       1275.2                 82               4   
3        1594       1275.2                 82               4   
4        1609       1287.2                 80               4   
...       ...          ...                ...             ...   
99995    1673       1338.4                 71               4   
99996    1667       1333.6                 71               4   
99997    1672       1337.6                 71               4   
99998    1686       1348.8                 69               4   
99999    1687       1349.6                 69               4   

       number_plan_changes  relocated  monthly_bill  \
0                        1          0           121   
1                        1          0           133   
2                        1          0           118   
3                        1          0           123   
4                        1          0           177   
...                    ...        ...           ...   
99995                    2          0            46   
99996                    2          0            47   
99997                    2          0            36   
99998                    2          0            47   
99999                    2          0            33   

       technical_issues_per_month  
0                               4  
1                               4  
2                               4  
3                               4  
4                               4  
...                           ...  
99995                           4  
99996                           4  
99997                           4  
99998                           4  
99999                           4  

[100000 rows x 8 columns]

In [11]:
Y=data.loc[:,["active_cust"]]
Y

active_cust
0                0
1                0
2                0
3                0
4                1
...            ...
99995            1
99996            1
99997            1
99998            1
99999            1

[100000 rows x 1 columns]

## Splitting the data into train and test 

## Building Models- Linear Regression

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [13]:
train_x,test_x,train_y,test_y=train_test_split(X,Y,test_size=.2,random_state=4)

In [14]:
lr=LinearRegression()

In [15]:
lr.fit(train_x,train_y)
y_pred=lr.predict(test_x)
y_pred

array([[0.3945226 ],
       [0.62263037],
       [0.63974256],
       ...,
       [0.55424792],
       [0.43407849],
       [0.55772371]])

In [16]:
accuracy = lr.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))

Accuracy = 0.24840130790954473


## KNN Regressor

In [17]:
from sklearn import neighbors
n_neighbors = 5
knn = neighbors.KNeighborsRegressor()
y_pred2 = knn.fit(train_x,train_y).predict(test_x)


In [18]:
accuracy = knn.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))

Accuracy = 0.21273779575461593


## Decision Tree Regressor

In [19]:
from sklearn.tree import DecisionTreeRegressor
regr_1 = DecisionTreeRegressor(max_depth=2)
regr_2 = DecisionTreeRegressor(max_depth=10)
regr_1.fit(train_x,train_y)
regr_2.fit(train_x,train_y)
y_pred3=regr_1.predict(test_x)
y_pred4=regr_2.predict(test_x)

In [20]:
accuracy = regr_1.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))
accuracy = regr_2.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))

Accuracy = 0.22127692713918323
Accuracy = 0.5260347510894382


## Random Forest Regressor

In [21]:
from sklearn.ensemble import RandomForestRegressor
rfr=RandomForestRegressor(n_estimators=100,max_depth=15)
rfr.fit(train_x,train_y)
y_pred5=rfr.predict(test_x)
accuracy = rfr.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))

C:\Users\Nikhitha Siva\Anaconda3\envs\new\lib\site-packages\ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


Accuracy = 0.5386234596717432


## Ada Boost Regressor

In [22]:
from sklearn.ensemble import AdaBoostRegressor
abr=AdaBoostRegressor(n_estimators=100)
abr.fit(train_x,train_y)
y_pred6=abr.predict(test_x)
accuracy = abr.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))

C:\Users\Nikhitha Siva\Anaconda3\envs\new\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Accuracy = 0.3205352058520383


## Tuning Parameters 

## Implementing GridSearch
#### When we compared all regression models, the Random Forest Regressor gave us better accuracy. Hence we are fine tuning the parameters using gridsearch in order to find out the best fit model.

In [23]:
from sklearn.model_selection import GridSearchCV
param_grid={'n_estimators':[150,200],'criterion':['mse'],'max_depth':[15,20,30]}
Gd=GridSearchCV(rfr,param_grid,cv=2,n_jobs=-1,verbose=2,return_train_score= True)

In [25]:
Gd.fit(X,Y)

Fitting 2 folds for each of 6 candidates, totalling 12 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 8 concurrent workers.
[Parallel(n_jobs=-1)]: Done   4 out of  12 | elapsed:  3.8min remaining:  7.5min
[Parallel(n_jobs=-1)]: Done  12 out of  12 | elapsed:  8.1min finished
C:\Users\Nikhitha Siva\Anaconda3\envs\new\lib\site-packages\sklearn\model_selection\_search.py:765: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  self.best_estimator_.fit(X, y, **fit_params)


GridSearchCV(cv=2, estimator=RandomForestRegressor(max_depth=15), n_jobs=-1,
             param_grid={'criterion': ['mse'], 'max_depth': [15, 20, 30],
                         'n_estimators': [150, 200]},
             return_train_score=True, verbose=2)

In [26]:
Gd.best_estimator_

RandomForestRegressor(max_depth=20, n_estimators=150)

In [27]:
Gd.best_score_

-0.4280798847344315

In [31]:
from sklearn.ensemble import RandomForestRegressor
rfr=RandomForestRegressor(n_estimators=150,max_depth=15)
rfr.fit(train_x,train_y.values.ravel())
y_pred7=rfr.predict(test_x)
accuracy = rfr.score(test_x, test_y)
print('Accuracy = '+ str(accuracy))

Accuracy = 0.5396497869078869


> The highest accuracy that was achieved is 53.9%

### Using more feature engineering and hyperparameter tuning the model can be further improved, but requires more time.
